### ISSUE TITLE
[Feature Request]: Integrate `gpuSort` into `sort`

### ISSUE THREAD

**Issue by ShreyasKhandekar** [chapel-lang/chapel#25363]
## Summary of Feature

**Description:**
Currently, users can call `gpuSort` from inside a `on here.gpus` block on an array accessible from GPU memory and it will sort the array on the GPU. This functionality has limitations today and will be expanded in later releases.

There was desire to integrate `gpuSort` into `sort` such that `sort` can intelligently use `gpuSort` when needed, first expressed in https://github.com/Cray/chapel-private/issues/4617#issuecomment-1775562049, but now this issue is making explicit note of it in a public issue.

### Code Sample

Here is how `gpuSort` should be integrated into `sort`.
```chapel
use Sort;
on here.gpus[0] {
  var A : [1..100] int;
  sort(A);  // Uses GPU based sort
}
var B : [1..100] int;
sort(B);  // Uses CPU based sort
```

## Stabilization concerns

Since `proc sort` is going to be stabilized, and `gpuSort` is far from being stable, the Sort module discussion subteam discussed this issue and decided the [following](https://github.com/chapel-lang/chapel/issues/24857#issuecomment-2190207121):

> - there are no changes we want to make to the API to support `gpuSort`
>    - sort should "just work", and whether it calls the existing implementation or `gpuSort` internally is an implementation detail/performance improvement.
>    - as far as API stability is concerned, there is nothing to do for now

**TL;DR:**
In essence, `gpuSort` is like an optimization that `sort` can use under the hood. And as long as correctness is guaranteed, we can expand `gpuSort` to take over "cpuSort"  in more and more cases as it gains functionality.

## Blockers for full integration

There are quite a few questions to be asked of which cases a GPU based sort should be used vs a CPU based sort.
Some of them are captured below, this is not an exhaustive list:

1. When we go back to the CPU after declaring a GPU array:
   ```chapel
   on gpu {
     var someGPUArray...
       on cpu {
         sort(someGPUArray)
      }
   }
   ```
2. Using the newly added remote variable declaration syntax (this is very similar to case 1 above)
    ```chapel
    on here.gpus[0] var remoteGPUArray...;
    sort(remoteGPUArray);
    ```
3. Sorting a CPU array in a GPU block:
    ```chapel
    var someCPUArray...
    on gpu {
      sort(someCPUArray)
    }
    ```
4. More weirdness with jumping between GPUs or GPUs across locales
    ```chapel
    var otherGpu = here.gpus[1];  // or otherLocale.gpus[1];
    on here.gpus[0]{
        var A = [(12.3, 3.2), (24, 4.7), (5.7, 4.2)];
        writeln(A);
        on otherGpu {
            sort(A);
        }
        writeln(A);
    }
    ```
Other blockers for complete integration:
 - `gpuSort` only takes arrays of numeric types that too not in all cases (ex: for ROCm version <4)
 - `gpuSort` only sorts in ascending order, and
 - `gpuSort`does not take comparators yet, and that may be a big lift since we'll have to teach CUB (the implementation used under the hood) how to interpret a Chapel comparator

None of these questions block the stabilization of `sort`. But are interesting to answer when we do try to integrate `gpuSort` into `sort`.

**Comment by e-kayrakli** [comment]
> In essence, gpuSort is like an optimization that sort can use under the hood. And as long as correctness is guaranteed, we can expand gpuSort to take over "cpuSort" in more and more cases as it gains functionality.

Agreed.

> There are quite a few questions to be asked of which cases a GPU based sort should be used vs a CPU based sort.
> Some of them are captured below, this is not an exhaustive list:

Ideal implementation of `sort` should use whichever locale(s) make sense to perform the operation. One could consider expanding the interface with optional arguments to limit the set of locales `sort` could use, but that's very future-looking. So my answers:

> 1. When we go back to the CPU after declaring a GPU array:

GPU

> 2. Using the newly added remote variable declaration syntax (this is very similar to case 1 above)

GPU (https://github.com/chapel-lang/chapel/issues/25351 is very relevant)

> 3. Sorting a CPU array in a GPU block:

CPU

> 4. More weirdness with jumping between GPUs or GPUs across locales

`Locales[0].gpus[0]`

> gpuSort only takes arrays of numeric types that too not in all cases (ex: for ROCm version <4)

If we were to proceed with integration with the capabilities we have today, we could fall back to copying back to host, sorting there and copying back, is that right? I don't think that's a showstopper. I recommend throwing in a no-doc `config param debugGpuSort` or something of that sort to allow us/users to understand the behavior easily.

> gpuSort only sorts in ascending order

This is probably a limitation on our side, right? I see [`SortKeysDescending`](https://nvidia.github.io/cccl/cub/api/structcub_1_1DeviceRadixSort.html#_CPPv4I00EN3cub15DeviceRadixSort18SortKeysDescendingE11cudaError_tPvR6size_tPK4KeyTP4KeyT9NumItemsTii12cudaStream_t) in CUB. It may also be relatively new, they seem to be moving CUB around quite a bit in the last couple of releases.

> gpuSortdoes not take comparators yet, and that may be a big lift since we'll have to teach CUB (the implementation used under the hood) how to interpret a Chapel comparator

I don't have anything to add. CPU fallback is the initial solution that we will live with until someone needs that operation to perform well, I think.
